src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSWebWorker.cs

```diff
@@ -30,220 +30,205 @@ public static Task RunAsync(Func<Task> body)
-            }
-            return await RunAsyncImpl(body, cancellationToken).ConfigureAwait(false);
+            var instance = new JSWebWorkerInstance<T>(body, null, cancellationToken);
+            return instance.Start();
         }
 
-        public static async Task RunAsync(Func<Task> body, CancellationToken cancellationToken)
+        public static Task RunAsync(Func<Task> body, CancellationToken cancellationToken)
         {
-            if (JSProxyContext.MainThreadContext.IsCurrentThread())
-            {
-                await JavaScriptImports.ThreadAvailable().ConfigureAwait(false);
-            }
-            await RunAsyncImpl(body, cancellationToken).ConfigureAwait(false);
+            var instance = new JSWebWorkerInstance<int>(null, body, cancellationToken);
+            return instance.Start();
         }
 
-        private static Task<T> RunAsyncImpl<T>(Func<Task<T>> body, CancellationToken cancellationToken)
+        internal sealed class JSWebWorkerInstance<T> : IDisposable
         {
-            var parentContext = SynchronizationContext.Current ?? new SynchronizationContext();
-            // continuation should not be running synchronously in the JSWebWorker thread because we are about to kill it after we resolve/reject the Task.
-            var tcs = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
-            var capturedContext = SynchronizationContext.Current;
-            var t = new Thread(() =>
-            {
-                try
+            private JSSynchronizationContext? _jsSynchronizationContext;
+            private TaskCompletionSource<T> _taskCompletionSource;
+            private Thread _thread;
+            private CancellationToken _cancellationToken;
+            private CancellationTokenRegistration? _cancellationRegistration;
+            private Func<Task<T>>? _bodyRes;
+            private Func<Task>? _bodyVoid;
+            private Task? _result;
+            private bool _isDisposed;
+
+            public JSWebWorkerInstance(Func<Task<T>>? bodyRes, Func<Task>? bodyVoid, CancellationToken cancellationToken)
+            {
+                _taskCompletionSource = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
+                _thread = new Thread(ThreadMain);
+                _result = null;
+                _cancellationToken = cancellationToken;
+                _cancellationRegistration = null;
+                _bodyRes = bodyRes;
+                _bodyVoid = bodyVoid;
+                JSHostImplementation.SetHasExternalEventLoop(_thread);
+            }
+
+            public Task<T> Start()
+            {
+                if (JSProxyContext.MainThreadContext.IsCurrentThread())
                 {
-                    if (cancellationToken.IsCancellationRequested)
+                    // give browser chance to load more threads
+                    JavaScriptImports.ThreadAvailable().ContinueWith(_ =>
```

Comment:
If we're at our limit will this just hang forever? Is there anything we can do about that failure mode if so?

Reply:
until there at least one thread loaded, it doesn't make sense to `Start`. Because that would also hang, but in a blocking way, much worse.

Reply:
I will add it to the comment

Reply:
Or do you mean we should give it something like 5 seconds and then throw ? 
I will try to at least respect cancelationToken